webui/react/src/pages/F_ExpList/glide-table/ColumnPickerMenu.tsx

```diff
@@ -63,31 +81,63 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
   }, [searchString, totalColumns, tab]);
 
   const allFilteredColumnsChecked = useMemo(() => {
-    return filteredColumns.map((col) => columnState.has(col.column)).every((col) => col === true);
+    return filteredColumns.every((col) => columnState.includes(col.column));
   }, [columnState, filteredColumns]);
 
   const handleShowHideAll = useCallback(() => {
     const filteredColumnMap: Record<string, boolean> = filteredColumns.reduce(
-      (acc, col) => ({ ...acc, [col.column]: columnState.has(col.column) }),
+      (acc, col) => ({ ...acc, [col.column]: columnState.includes(col.column) }),
       {},
     );
 
     const newColumns = allFilteredColumnsChecked
-      ? [...columnState].filter((col) => !filteredColumnMap[col])
+      ? columnState.filter((col) => !filteredColumnMap[col])
       : [...new Set([...columnState, ...filteredColumns.map((col) => col.column)])];
     setVisibleColumns(newColumns);
-  }, [allFilteredColumnsChecked, filteredColumns, setVisibleColumns, columnState]);
+
+    // If uncheck something pinned, reduce the pinnedColumnsCount
+    allFilteredColumnsChecked &&
+      updateSettings({
+        pinnedColumnsCount: newColumns.filter(
+          (col) => columnState.indexOf(col) < settings.pinnedColumnsCount,
+        ).length,
+      });
+  }, [
+    allFilteredColumnsChecked,
+    filteredColumns,
+    setVisibleColumns,
+    columnState,
+    updateSettings,
+    settings.pinnedColumnsCount,
+  ]);
 
   const handleColumnChange = useCallback(
     (event: CheckboxChangeEvent) => {
       const { id, checked } = event.target;
       if (id === undefined) return;
-
-      const newColumnSet = new Set(columnState);
-      checked ? newColumnSet.add(id) : newColumnSet.delete(id);
-      setVisibleColumns([...newColumnSet]);
+      const pinnedColumnsCount = settings.pinnedColumnsCount;
+      if (settings.compare) {
+        // pin or unpin column
+        const newColumns = columnState.filter((c) => c !== id);
+        if (checked) {
+          newColumns.splice(pinnedColumnsCount, 0, id);
+          updateSettings({ pinnedColumnsCount: Math.max(pinnedColumnsCount + 1, 0) });
+        } else {
+          newColumns.splice(pinnedColumnsCount - 1, 0, id);
+          updateSettings({ pinnedColumnsCount: Math.max(pinnedColumnsCount - 1, 0) });
+        }
+        setVisibleColumns(newColumns);
+      } else {
+        // If uncheck something pinned, reduce the pinnedColumnsCount
+        if (!checked && columnState.indexOf(id) < pinnedColumnsCount) {
+          updateSettings({ pinnedColumnsCount: Math.max(pinnedColumnsCount - 1, 0) });
+        }
+        const newColumnSet = new Set(columnState);
+        checked ? newColumnSet.add(id) : newColumnSet.delete(id);
```

Comment:
nit: i think `if statement` is more meaningful than ternary operator here even though its a bit redundant because ternary operator is mostly used to return value, but this line doesn't return a value 

```suggestion
        if (checked) {
          newColumnSet.add(id);
        } else {
          newColumnSet.delete(id);
        }
```